src/libraries/Microsoft.Extensions.Options/src/OptionsCache.cs

```diff
@@ -45,6 +45,29 @@ public virtual TOptions GetOrAdd(string? name, Func<TOptions> createOptions!!)
             return value.Value;
         }
 
+        internal TOptions GetOrAdd<TArg>(string? name, Func<string, TArg, TOptions> createOptions, TArg factoryArgument)
+        {
+            // For compatibility, fall back to public GetOrAdd() if we're in a derived class.
+            if (GetType() != typeof(OptionsCache<TOptions>))
+            {
+                return GetOrAdd(name, () => createOptions(name ?? Options.DefaultName, factoryArgument));
```

Comment:
Note that even though this is guarded behind an if, it's still going to force a closure allocation at the beginning of the method for all uses due to the createOptions parameter being captured.

Reply:
I added a method to create indirection. Is that sufficient to fix?

Reply:
You can do that, or you can just create a local inside the scope where you want the closure allocated and capture that local, e.g.
```C#
if (GetType() != typeof(OptionsCache<TOptions>))
{
    Func<TOptions> localCreateOptions = createOptions;
    TArg localFactoryArgument = factoryArgument;
    return GetOrAdd(name, () => localCreateOptions(name ?? Options.DefaultName, localFactoryArgument));
}
```

Reply:
Makes sense. Which pattern do you prefer here?

Reply:
I don't have a strong preference.  We've typically done the locals thing, though it's easier to "get it right" with a separate method.

Reply:
It looks like the legacy framework branches should have the same issue because of the delegate passed to the Lazy constructor in the cache miss case. Do you think it is worth fixing for those as well?